### DISCUSSION TITLE
CLOG 1.4 - Dev Journal

### DISCUSSION THREAD

**Discussion by rabbibotton** [rabbibotton/clog#149] [General]
The main goal of CLOG 1.4 is to improve the builders ability to layout controls, improve interface and overall speed.

**Comment by rabbibotton** [comment]
Now moving and resizing dynamically resizes controls not just a bounding box
Control "snap" around and to each other

**Comment by rabbibotton** [comment]
No longer has excessive repolling of properties resulting in incredible speed difference for using the buidler

**Comment by rabbibotton** [comment]
support on-load property for control-records for reattaching custom code after page load and support for composite controls

**Comment by rabbibotton** [comment]
New feature that is really cool :) - if permissions granted to clog builder then will use the system clip board. This means you can copy items to the clip board, paste in a text editor and modify the HTML and paste back. You can also just select html and from any place and paste it in to the builder - make sure it is wrapped in a div ideally.

**Reply by rabbibotton** [reply]
Implemented keyboard and also _browser_ menu support for cut/copy/paste cmd/crl-x/c/v

**Comment by rabbibotton** [comment]
Icon bar :)

![Screen Shot 2022-06-24 at 12 05 38 PM](https://user-images.githubusercontent.com/11616789/175574243-6a705cfc-de1c-468e-b83b-cf658317c7aa.png)

**Comment by rabbibotton** [comment]
More work towards new look and feel. The right panel can be shut and open clicking on black bar.

<img width="1066" alt="Screen Shot 2022-06-24 at 6 20 38 PM" src="https://user-images.githubusercontent.com/11616789/175701360-00328ab2-d1ca-4681-bdfd-ddaee2688409.png">

**Comment by rabbibotton** [comment]
Little more work on style :)

<img width="1234" alt="clog-builder" src="https://user-images.githubusercontent.com/11616789/175799848-c73788ea-d960-436a-b6f6-d9d3f1bedf24.png">

**Comment by rabbibotton** [comment]
You can now use keyboard arrows to move around controls, shift-and arrows to resize

**Comment by rabbibotton** [comment]
added shift+mouse click to make target control a child of current control

**Comment by rabbibotton** [comment]
you can now drop directly on other controls, if hit shift become children
if ctrl/command static placed
double click on control makes it unreachable so can adjust children

**Comment by rabbibotton** [comment]
Added to all common elements and form elements a :style option for hand coding ease.

**Comment by rabbibotton** [comment]
Rendered builder code will now keep the data-clog-name attribute. This way when needed you can use the control names set in the builder with JavaScript and other code since the html id changes and can not be used:

(js-execute target "$('[data-clog-name=\'button-1\']').css('color','red')")

**Reply by rabbibotton** [reply]
with that change - the labels "for" property can now be set using the control's name and persist in saves and of course in renders.

**Comment by rabbibotton** [comment]
Undo and Redo implemented

**Reply by aykaramba** [reply]
Amazing!

**Reply by rabbibotton** [reply]
For sure is, if you have a bunch of controls embedded in a div and a little slip of your finger all gone... So this was a must :)

**Comment by rabbibotton** [comment]
New tool : image-to-data in the tools menu

Convert an image to embedded data that can be place directly in your Lisp source or html files or even pasted in to a CLOG Builder property such as src-url on an img. This removes need to distribute an image file.

![Screen Shot 2022-06-30 at 3 01 51 PM](https://user-images.githubusercontent.com/11616789/176757912-7c74c70d-80ae-44d4-a2b9-4a8c099a8361.png)

**Comment by rabbibotton** [comment]
tool tips on menu bars

**Reply by aykaramba** [reply]
Amazing.

**Comment by rabbibotton** [comment]
alignment boxes

**Comment by rabbibotton** [comment]
clog-gui's server-file-dialog works properly now when initial-path has a filename and set builder to use decent defaults

**Comment by rabbibotton** [comment]
Size adjustable palette vs control list panes

**Comment by rabbibotton** [comment]
copy/cut history window, any copy or cut made in builder saved to a buffer of entire session and accessible in the copy history window.

**Comment by rabbibotton** [comment]
added quick start window to the help
